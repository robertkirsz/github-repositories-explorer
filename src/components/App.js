import Div from 'styled-kit/Div'
import UsersList from 'components/UsersList'
import SearchInfo from 'components/SearchInfo'
import SearchMoreButton from 'components/SearchMoreButton'
import SearchUsersInput from 'components/SearchUsersInput'
import ErrorMessage from 'components/ErrorMessage'

export default function App() {
  return (
    <Div column listTop width="100%" maxWidth={480} mobile="overflow: hidden;">
      <h1 className="title">Repos Explorer</h1>
      <SearchUsersInput />
      <SearchInfo />
      <UsersList />
      <SearchMoreButton />
      <ErrorMessage />
    </Div>
  )
}
