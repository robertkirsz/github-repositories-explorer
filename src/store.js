import { configureStore } from '@reduxjs/toolkit'
import rootReducers from 'reducers'

const store = configureStore({ reducer: rootReducers })

// Reuse the store if any reducer is recompiled during hot-reloading
if (process.env.NODE_ENV === 'development' && module.hot) {
  module.hot.accept('reducers', () => {
    const newRootReducer = require('reducers').default
    store.replaceReducer(newRootReducer)
  })
}

export default store
