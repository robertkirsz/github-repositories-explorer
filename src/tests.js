import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import store from 'store'
import { GlobalStyle } from 'styles'
import '@testing-library/jest-dom/extend-expect'
import 'jest-styled-components'

const AllTheProviders = ({ children = null }) => (
  <Provider store={store}>
    {children}
    <GlobalStyle />
  </Provider>
)

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
