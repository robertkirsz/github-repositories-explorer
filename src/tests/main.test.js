import { rest } from 'msw'
import { setupServer } from 'msw/node'
import { render, fireEvent } from 'tests/test-utils'
import '@testing-library/jest-dom/extend-expect'
import store from 'store'
import App from 'components/App'

const server = setupServer(
  rest.get('https://api.github.com/search/users', (request, response, context) => {
    const pages = {
      1: [
        {
          id: 1,
          login: 'john',
          repos_url: 'https://api.github.com/users/john/repos'
        },
        {
          id: 2,
          login: 'johnny',
          repos_url: 'https://api.github.com/users/johnny/repos'
        }
      ],
      2: [
        {
          id: 3,
          login: 'jonnie',
          repos_url: 'https://api.github.com/users/jonnie/repos'
        }
      ]
    }

    return response(context.json({ items: pages[request.url.searchParams.get('page')] }))
  }),
  rest.get('https://api.github.com/users/:username/repos', (_, response, context) =>
    response(
      context.json([
        {
          id: 1,
          name: "John's repo",
          description: "John's first project",
          stargazers_count: 2,
          html_url: 'https://github.com/john/johns-repo'
        }
      ])
    )
  )
)

test('The flow works', async () => {
  // Start server
  server.listen()

  // Render the app
  const { container, getByText, findByText, getByPlaceholderText } = render(<App />)

  // Title should be visible from the start
  expect(getByText('GitHub Repositories Explorer')).toBeVisible()

  // Type "john" into the text field
  fireEvent.change(getByPlaceholderText('Enter username'), { target: { value: 'john' } })

  // Loader should appear
  expect(await findByText('Loading...')).toBeVisible()
  expect(store.getState().users.lastSearchedUsername).toEqual('john')

  // Info label should appear after we get 'fetchUsers' API response
  expect(await findByText('Showing users for "john"')).toBeVisible()
  expect(store.getState().users.items).toHaveLength(2)

  // These two users should be visible
  expect(getByText('john')).toBeVisible()
  expect(getByText('johnny')).toBeVisible()

  // Click the "Search more" button
  fireEvent.click(getByText('Search more'))

  // Wait for another 'fetchUsers' API response
  expect(await findByText('Loading...')).toBeVisible()
  expect(store.getState().users.currentPage).toEqual(2)

  // Third user should appear
  expect(await findByText('jonnie')).toBeVisible()
  expect(store.getState().users.items).toHaveLength(3)

  // Click on the first user
  fireEvent.click(getByText('john'))

  // Repository info should appear after we get 'fetchUserRepos' API response
  expect(await findByText('Loading...')).toBeVisible()
  expect(await findByText("John's repo")).toBeVisible()
  expect(store.getState().users.items[0].repos).toHaveLength(1)

  // Make sure the UI didn't change
  expect(container).toMatchSnapshot()

  // Stop server
  server.resetHandlers()
  server.close()
})
