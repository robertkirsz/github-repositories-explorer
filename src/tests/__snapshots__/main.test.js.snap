// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The flow works 1`] = `
<div>
  <div
    class="sc-1qr1lby-0 App___StyledDiv-sc-1ruxc1o-0 gHPidm fQgRfj"
    width="100%"
  >
    <h1
      class="App___StyledH-sc-1ruxc1o-1 ela-dlJ"
    >
      GitHub Repositories Explorer
    </h1>
    <input
      class="DebouncedInput__Input-sc-13ol9vv-0 cnLTLN"
      placeholder="Enter username"
      value="john"
    />
    <div
      class="InfoMessage-sc-4uxl10-0 gIZCtt"
    >
      Showing users for "john"
    </div>
    <div
      class="sc-1qr1lby-0 UsersList___StyledDiv-sc-1w8kqhz-0 igLjTD eIGSlG"
    >
      <div
        class="sc-1qr1lby-0 bBOvBH"
      >
        <div
          class="sc-1qr1lby-0 jRuxNE"
          radius="3"
        >
          <span
            class="truncated"
          >
            john
          </span>
          <span
            class="Chevron-sc-1w1czyc-0 fuWAsm"
          />
        </div>
        <div
          class="sc-1qr1lby-0 ReposList___StyledDiv-sc-1ucb7ap-0 dPhwYJ imZKwy"
        >
          <div
            class="sc-1qr1lby-0 fisVz"
            radius="3"
          >
            <div
              class="sc-1qr1lby-0 ctapDY"
              font-weight="600"
            >
              <a
                as="a"
                href="https://github.com/john/johns-repo"
                rel="noreferrer"
                target="_blank"
              >
                John's repo
              </a>
              <span>
                2
                 
                <span
                  aria-hidden="false"
                  aria-label="Star emoji"
                  class="Emoji-sc-1mu2kvi-0 kHwBut"
                  role="img"
                >
                  ⭐️
                </span>
              </span>
            </div>
            <span>
              John's first project
            </span>
          </div>
        </div>
      </div>
      <div
        class="sc-1qr1lby-0 bBOvBH"
      >
        <div
          class="sc-1qr1lby-0 jRuxNE"
          radius="3"
        >
          <span
            class="truncated"
          >
            johnny
          </span>
          <span
            class="Chevron-sc-1w1czyc-0 eiiklV"
          />
        </div>
      </div>
      <div
        class="sc-1qr1lby-0 bBOvBH"
      >
        <div
          class="sc-1qr1lby-0 jRuxNE"
          radius="3"
        >
          <span
            class="truncated"
          >
            jonnie
          </span>
          <span
            class="Chevron-sc-1w1czyc-0 eiiklV"
          />
        </div>
      </div>
    </div>
    <button
      class="SearchMoreButton___StyledButton-ec5qj5-0 gldqLk"
    >
      Search more
    </button>
  </div>
</div>
`;
